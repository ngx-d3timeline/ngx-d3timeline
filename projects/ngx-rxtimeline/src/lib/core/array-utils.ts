export function sum(...args: number[]) {
  return args.reduce((a, b) => a + b);
}

export function toArray(...args: any[]): any[] {
  return args;
}
