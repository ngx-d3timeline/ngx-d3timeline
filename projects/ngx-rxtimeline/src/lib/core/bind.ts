export function bind1<TBoundArg, TUnboundArgs extends any[], TOut>(
  func: (boundArg: TBoundArg, ...unboundArgs: TUnboundArgs) => TOut,
  boundVal: TBoundArg
): (...unboundArgs: TUnboundArgs) => TOut {
  return (...unboundArgs: TUnboundArgs) => func(boundVal, ...unboundArgs);
}

export function bind2<TBoundArg1, TBoundArg2, TUnboundArgs extends any[], TOut>(
  func: (
    boundArg1: TBoundArg1,
    boundArg2: TBoundArg2,
    ...unboundArgs: TUnboundArgs
  ) => TOut,
  boundVal1: TBoundArg1,
  boundVal2: TBoundArg2
): (...unboundArgs: TUnboundArgs) => TOut {
  return (...unboundArgs: TUnboundArgs) =>
    func(boundVal1, boundVal2, ...unboundArgs);
}
