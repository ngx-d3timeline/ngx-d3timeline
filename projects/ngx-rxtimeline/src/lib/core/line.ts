import { Orientation } from './orientation';
import { Point, translatePointInOrientation, origin } from './point';
import { partial } from './partial';

export interface Line {
  x1: number;
  x2: number;
  y1: number;
  y2: number;
}

export const createOrientedLineFromOrigin = partial(createOrientedLine, origin);

export function createOrientedLine(
  point: Point,
  length: number,
  orientation: Orientation
): Line {
  const to = translatePointInOrientation(point, length, orientation);

  return createLine(point, to);
}

export function createLine(from: Point, to: Point) {
  return {
    x1: from.x,
    y1: from.y,
    x2: to.x,
    y2: to.y
  };
}
