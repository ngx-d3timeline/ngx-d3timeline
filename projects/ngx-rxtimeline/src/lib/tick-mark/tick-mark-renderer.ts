import { Orientation } from '../core/orientation';

export interface TickMarkRenderer {
  orientation: Orientation;
  mapTickValueToPositionInScale(tickValue: any): number;
}
