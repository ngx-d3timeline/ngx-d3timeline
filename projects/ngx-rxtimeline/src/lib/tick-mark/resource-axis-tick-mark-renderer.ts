import { BandScale } from '../scales/scale-types';
import { TickMarkRenderer } from '../tick-mark/tick-mark-renderer';
import { OrientedScale } from '../scales/oriented-scale';

export function getResourceAxisTickMarkRenderer(
  orientedScale: OrientedScale<BandScale>
): TickMarkRenderer {
  function getBandMidPoint(tickValue: string): number {
    return orientedScale.scale(tickValue) + orientedScale.scale.bandwidth() / 2;
  }

  return {
    orientation: orientedScale.orientation,
    mapTickValueToPositionInScale: (tickValue: string) =>
      getBandMidPoint(tickValue)
  };
}
