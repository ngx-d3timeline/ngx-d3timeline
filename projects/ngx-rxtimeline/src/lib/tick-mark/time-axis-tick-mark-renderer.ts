import { TimeScale } from '../scales/scale-types';
import { TickMarkRenderer } from './tick-mark-renderer';
import { OrientedScale } from '../scales/oriented-scale';

export function getTimeAxisTickMarkRenderer(
  orientedScale: OrientedScale<TimeScale>
): TickMarkRenderer {
  return {
    orientation: orientedScale.orientation,
    mapTickValueToPositionInScale: (value: Date) => orientedScale.scale(value)
  };
}
