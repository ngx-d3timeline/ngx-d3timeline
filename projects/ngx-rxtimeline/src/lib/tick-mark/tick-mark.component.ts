import { Component, Input } from '@angular/core';
import { TickMark } from './tick-mark';

@Component({
  selector: '[ngx-rxtimeline-tick-mark]',
  template: `
    <svg:g [attr.transform]="tickMark.transform">
      <svg:text
        [attr.dx]="tickMark.labelOffset.x"
        [attr.dy]="tickMark.labelOffset.y"
        [attr.font-family]="tickMark.fontFace"
        [attr.font-size]="tickMark.fontSize"
        class="tick-mark-text"
      >
        {{ tickMark.label }}
      </svg:text>
      <svg:g
        class="tick-mark-line"
        ngx-rxtimeline-line
        *ngIf="tickMark.line"
        [line]="tickMark.line"
      ></svg:g>
    </svg:g>
  `
})
export class TickMarkComponent {
  @Input() tickMark: TickMark;
}
