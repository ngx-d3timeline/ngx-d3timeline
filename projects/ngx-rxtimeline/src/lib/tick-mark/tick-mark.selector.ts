import { createSelector } from '../store-lib/selector/create-selector';

import { selectResources } from '../scales/selectors/band-scale.selectors';
import { selectOrientedScale } from '../scales/selectors/scale-selectors';
import { selectTimeScale } from '../scales/selectors/time-scale.selectors';
import { getTimeAxisTickMarkRenderer } from './time-axis-tick-mark-renderer';
import { getResourceAxisTickMarkRenderer } from './resource-axis-tick-mark-renderer';
import { mapValues } from '../core/transform-utils';
import { selectViewTopLeft } from '../view/view.selectors';
import {
  selectAxisTickLineOffset,
  selectAxisFontFace,
  selectAxisFontSize
} from '../options/selectors/axis-options.selectors';
import { partialApply } from '../core/function-utils';
import { TimeScale } from '../scales/scale-types';
import { Orientation, flipOrientation } from '../core/orientation';
import { Point, pointToTransform, origin } from '../core/point';
import { TickMarkRenderer } from './tick-mark-renderer';
import { TickMark } from './tick-mark';
import { createOrientedLine } from '../core/line';
import { AxisType } from '../axis/axis';
import { createEnumSelector } from '../store-lib/selector/selector-utils';
import { constSelector } from '../store-lib/selector/selector';

export const selectGetTickPosition = createSelector(
  selectViewTopLeft,
  viewTopLeft => getTickPosition.bind(null, viewTopLeft)
);

function getTickPosition(
  viewTopLeft: Point,
  orientation: Orientation,
  range: number
): Point {
  return orientation === Orientation.Vertical
    ? { ...viewTopLeft, y: range }
    : { ...viewTopLeft, x: range };
}

const selectResourceAxisTickMarkRenderer = createSelector(
  selectOrientedScale(AxisType.Resources),
  selectAxisTickLineOffset(AxisType.Resources),
  getResourceAxisTickMarkRenderer
);

const selectTimeAxisTickMarkRenderer = createSelector(
  selectOrientedScale(AxisType.Time),
  selectAxisTickLineOffset(AxisType.Time),
  getTimeAxisTickMarkRenderer
);

const selectGetResourceAxisTickMark = createSelector(
  selectGetTickPosition,
  selectResourceAxisTickMarkRenderer,
  selectAxisFontFace(AxisType.Resources),
  selectAxisFontSize(AxisType.Resources),
  partialApply(getTickMark)
);

const selectGetTimeAxisTickMark = createSelector(
  selectGetTickPosition,
  selectTimeAxisTickMarkRenderer,
  selectAxisFontFace(AxisType.Time),
  selectAxisFontSize(AxisType.Time),
  partialApply(getTickMark)
);

function getTickMark(
  tickValue: any,
  tickPosition: (o: Orientation, range: number) => Point,
  tickMarkRenderer: TickMarkRenderer,
  fontFace: string,
  fontSize: number
): TickMark {
  return {
    label: tickMarkRenderer.getTickLabel(tickValue),
    transform: pointToTransform(
      tickPosition(
        tickMarkRenderer.orientation,
        tickMarkRenderer.mapTickValueToPositionInScale(tickValue)
      )
    ),
    labelOffset: getTickLabelOffset(
      tickMarkRenderer.getTickLabelSpacing(),
      flipOrientation(tickMarkRenderer.orientation)
    ),
    line: getTickLine(
      tickMarkRenderer.tickLineOffset,
      flipOrientation(tickMarkRenderer.orientation)
    ),
    fontFace,
    fontSize
  };
}

function getTickLine(lineOffset: number, orientation: Orientation) {
  return lineOffset && createOrientedLine(origin, lineOffset, orientation);
}

function getTickLabelOffset(labelSpacing: number, orientation: Orientation) {
  return orientation === Orientation.Vertical
    ? { ...origin, y: labelSpacing }
    : { ...origin, x: labelSpacing };
}

const selectTimeAxisTickValues = createSelector(
  selectTimeScale,
  getTimeAxisTickValues
);

function getTimeAxisTickValues(scale: TimeScale) {
  return scale.ticks();
}

export const selectAxisTickValues = (axisType: AxisType) =>
  createEnumSelector<AxisType, any[]>({
    Resources: selectResources,
    Time: selectTimeAxisTickValues
  })(constSelector(axisType));

export const selectResourceAxisTickMarks = createSelector(
  selectAxisTickValues(AxisType.Resources),
  selectGetResourceAxisTickMark,
  mapValues
);

export const selectTimeAxisTickMarks = createSelector(
  selectAxisTickValues(AxisType.Time),
  selectGetTimeAxisTickMark,
  mapValues
);
