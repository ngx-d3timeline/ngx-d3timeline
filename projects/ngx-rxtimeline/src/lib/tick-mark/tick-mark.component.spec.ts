import { TestBed, ComponentFixture } from '@angular/core/testing';
import { TickMarkComponent } from './tick-mark.component';
import { createLine, Line } from '../core/line';
import { Component, Input } from '@angular/core';

@Component({
  selector: '[ngx-rxtimeline-line]',
  template: `
    <svg:g></svg:g>
  `
})
class FakeLineComponent {
  @Input() line: Line;
}

describe('AxisTickMarkComponent', () => {
  let fixture: ComponentFixture<TickMarkComponent>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [TickMarkComponent, FakeLineComponent]
    });

    fixture = TestBed.createComponent(TickMarkComponent);
  });

  it('should not render line when tickMark.line is falsy', () => {
    fixture.componentInstance.tickMark = {
      label: 'tickMark label',
      labelOffset: { x: 0, y: 1 },
      line: null,
      transform: 'translate(10,10)',
      fontFace: 'arial',
      fontSize: 10
    };

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });

  it('should render line when tickMark.line is truthy', () => {
    fixture.componentInstance.tickMark = {
      label: 'tickMark label',
      labelOffset: { x: 0, y: 2 },
      line: createLine({ x: 0, y: 0 }, { x: 10, y: 10 }),
      transform: 'translate(10,10)',
      fontFace: 'arial',
      fontSize: 10
    };

    fixture.detectChanges();

    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
