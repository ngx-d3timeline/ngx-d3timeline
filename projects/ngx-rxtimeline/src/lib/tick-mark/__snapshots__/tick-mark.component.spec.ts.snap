// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AxisTickMarkComponent should not render line when tickMark.line is falsy 1`] = `
<div
  id="root0"
  ngx-rxtimeline-tick-mark=""
>
  <g
    transform="translate(10,10)"
  >
    <text
      class="tick-mark-text"
      dx="0"
      dy="1"
      font-family="arial"
      font-size="10"
    >
       tickMark label 
    </text>
    
  </g>
</div>
`;

exports[`AxisTickMarkComponent should render line when tickMark.line is truthy 1`] = `
<div
  id="root1"
  ngx-rxtimeline-tick-mark=""
>
  <g
    transform="translate(10,10)"
  >
    <text
      class="tick-mark-text"
      dx="0"
      dy="2"
      font-family="arial"
      font-size="10"
    >
       tickMark label 
    </text>
    
    <g
      class="tick-mark-line"
      ngx-rxtimeline-line=""
    >
      <g />
    </g>
  </g>
</div>
`;
